File: v4/rust/iprange-livedb/src/recovery/membership_table.rs
//! Fixed-record membership locator storage and ID lookup.

use crate::contract::{u16_le, u32_le};
use crate::error::{Error, Result};
use crate::membership_dictionary::codec::Storage;

pub(crate) use super::id_table::Insert;
use super::id_table::{IdIndex, RecordCodec};
use super::membership_index::Locator;
use super::tables::{Layout, Region, MEMBERSHIP_RECORD_SIZE};

const RECORD_SIZE: usize = MEMBERSHIP_RECORD_SIZE as usize;
const RECORD_PRESENT_OFFSET: usize = 0;
const RECORD_STORAGE_OFFSET: usize = 1;
const RECORD_LEAF_INDEX_OFFSET: usize = 2;
const RECORD_ID_OFFSET: usize = 4;
const RECORD_WORD_COUNT_OFFSET: usize = 8;
const RECORD_LEAF_PAGE_OFFSET: usize = 12;
const RECORD_BLOB_ROOT_OFFSET: usize = 16;
const RECORD_REJECTED_OFFSET: usize = 20;
const RECORD_DIGEST_OFFSET: usize = 24;
const RECORD_DIGEST_END: usize = 56;

pub(crate) type MembershipIndex = IdIndex<MembershipCodec>;

pub(crate) struct MembershipCodec;

impl RecordCodec for MembershipCodec {
    type Record = Locator;

    const WIDTH: usize = RECORD_SIZE;
    const INVALID_RECORD: &'static str = "recovery membership record index is invalid";
    const FULL: &'static str = "recovery membership ID table is full";

    fn regions(layout: Layout) -> (Region, Region) {
        let Layout {
            membership_records,
            membership_ids,
            ..
        } = layout;
        (membership_records, membership_ids)
    }

    fn encode(record: Locator, output: &mut [u8]) {
        encode(record, output)
    }

    fn decode(input: &[u8]) -> Result<Locator> {
        decode(input)
    }

    fn rejected(record: Locator) -> bool {
        record.rejected
    }

    fn reject(record: &mut Locator) {
        record.rejected = true;
    }
}

fn encode(locator: Locator, output: &mut [u8]) {
    debug_assert_eq!(output.len(), RECORD_SIZE);
    output.fill(0);
    output[RECORD_PRESENT_OFFSET] = 1;
    let root = match locator.storage {
        Storage::Inline => 0,
        Storage::Blob(root) => {
            output[RECORD_STORAGE_OFFSET] = 1;
            root
        }
    };
    output[RECORD_LEAF_INDEX_OFFSET..RECORD_ID_OFFSET]
        .copy_from_slice(&locator.leaf_index.to_le_bytes());
    output[RECORD_ID_OFFSET..RECORD_WORD_COUNT_OFFSET].copy_from_slice(&locator.id.to_le_bytes());
    output[RECORD_WORD_COUNT_OFFSET..RECORD_LEAF_PAGE_OFFSET]
        .copy_from_slice(&locator.word_count.to_le_bytes());
    output[RECORD_LEAF_PAGE_OFFSET..RECORD_BLOB_ROOT_OFFSET]
        .copy_from_slice(&locator.leaf_page.to_le_bytes());
    output[RECORD_BLOB_ROOT_OFFSET..RECORD_REJECTED_OFFSET].copy_from_slice(&root.to_le_bytes());
    output[RECORD_REJECTED_OFFSET] = u8::from(locator.rejected);
    output[RECORD_DIGEST_OFFSET..RECORD_DIGEST_END].copy_from_slice(&locator.digest);
}

fn decode(bytes: &[u8]) -> Result<Locator> {
    if bytes.len() != RECORD_SIZE {
        return Err(Error::Corrupt("recovery membership locator has wrong size"));
    }
    if bytes[RECORD_PRESENT_OFFSET] != 1 || bytes[RECORD_STORAGE_OFFSET] > 1 {
        return Err(Error::Corrupt("recovery membership locator is malformed"));
    }
    let root = u32_le(bytes, RECORD_BLOB_ROOT_OFFSET);
    let storage = match (bytes[RECORD_STORAGE_OFFSET], root) {
        (0, 0) => Storage::Inline,
        (1, root) if root >= 2 => Storage::Blob(root),
        _ => {
            return Err(Error::Corrupt(
                "recovery membership locator storage is malformed",
            ))
        }
    };
    let mut digest = [0; 32];
    digest.copy_from_slice(&bytes[RECORD_DIGEST_OFFSET..RECORD_DIGEST_END]);
    Ok(Locator {
        id: u32_le(bytes, RECORD_ID_OFFSET),
        word_count: u32_le(bytes, RECORD_WORD_COUNT_OFFSET),
        digest,
        leaf_page: u32_le(bytes, RECORD_LEAF_PAGE_OFFSET),
        leaf_index: u16_le(bytes, RECORD_LEAF_INDEX_OFFSET),
        storage,
        rejected: bytes[RECORD_REJECTED_OFFSET] != 0,
    })
}

File: v4/go/internal/recovery/membership_table.go
package recovery

// Membership locator table (Rust recovery/membership_table.rs): the
// fixed 56-byte locator records of the recovered membership
// dictionary, with the exact wire offsets of the Rust records.

import "github.com/firehol/iprange/v4/go/internal/format"

const (
	membershipRecordPresentOffset = 0
	membershipRecordStorageOffset = 1
	membershipRecordLeafIndexOff  = 2
	membershipRecordIDOffset      = 4
	membershipRecordWordCountOff  = 8
	membershipRecordLeafPageOff   = 12
	membershipRecordBlobRootOff   = 16
	membershipRecordRejectedOff   = 20
	membershipRecordDigestOffset  = 24
	membershipRecordDigestEnd     = 56
)

// membershipLocator is one recovered membership dictionary record
// (Rust recovery membership_index::Locator): the source facts needed
// to verify and rebuild one bitmap.
type membershipLocator struct {
	id        uint32
	wordCount uint32
	digest    [32]byte
	leafPage  uint32
	leafIndex uint16
	blobRoot  uint32
	storage   format.MembershipStorage
	rejected  bool
}

// membershipCodec adapts the locator to the shared id table (Rust
// MembershipCodec).
func membershipCodec() recordCodec[membershipLocator] {
	return recordCodec[membershipLocator]{
		width:         int(membershipRecordSize),
		invalidRecord: "recovery membership record index is invalid",
		full:          "recovery membership ID table is full",
		regions: func(layout tableLayout) (tableRegion, tableRegion) {
			return layout.membershipRecords, layout.membershipIDs
		},
		encode:     encodeMembershipLocator,
		decode:     decodeMembershipLocator,
		isRejected: func(record membershipLocator) bool { return record.rejected },
		reject:     func(record *membershipLocator) { record.rejected = true },
	}
}

// encodeMembershipLocator encodes one locator (Rust membership_table::
// encode: the present flag, the storage flag with the inline root
// zero, and every wire field).
func encodeMembershipLocator(locator membershipLocator, output []byte) {
	for index := range output {
		output[index] = 0
	}
	output[membershipRecordPresentOffset] = 1
	root := uint32(0)
	if locator.storage == format.MembershipStorageBlob {
		output[membershipRecordStorageOffset] = 1
		root = locator.blobRoot
	}
	format.PutU16(output[membershipRecordLeafIndexOff:membershipRecordIDOffset], locator.leafIndex)
	format.PutU32(output[membershipRecordIDOffset:membershipRecordWordCountOff], locator.id)
	format.PutU32(output[membershipRecordWordCountOff:membershipRecordLeafPageOff], locator.wordCount)
	format.PutU32(output[membershipRecordLeafPageOff:membershipRecordBlobRootOff], locator.leafPage)
	format.PutU32(output[membershipRecordBlobRootOff:membershipRecordRejectedOff], root)
	output[membershipRecordRejectedOff] = 0
	if locator.rejected {
		output[membershipRecordRejectedOff] = 1
	}
	copy(output[membershipRecordDigestOffset:membershipRecordDigestEnd], locator.digest[:])
}

// decodeMembershipLocator decodes one locator (Rust membership_table::
// decode: the present flag, the storage proof, and every wire field;
// every refusal is the Corrupt class).
func decodeMembershipLocator(bytes []byte) (membershipLocator, error) {
	if len(bytes) != int(membershipRecordSize) {
		return membershipLocator{}, corruptError("recovery membership locator has wrong size")
	}
	if bytes[membershipRecordPresentOffset] != 1 || bytes[membershipRecordStorageOffset] > 1 {
		return membershipLocator{}, corruptError("recovery membership locator is malformed")
	}
	root := format.U32(bytes[membershipRecordBlobRootOff:membershipRecordRejectedOff])
	var storage format.MembershipStorage
	switch {
	case bytes[membershipRecordStorageOffset] == 0 && root == 0:
		storage = format.MembershipStorageInline
	case bytes[membershipRecordStorageOffset] == 1 && root >= 2:
		storage = format.MembershipStorageBlob
	default:
		return membershipLocator{}, corruptError("recovery membership locator storage is malformed")
	}
	var locator membershipLocator
	copy(locator.digest[:], bytes[membershipRecordDigestOffset:membershipRecordDigestEnd])
	locator.id = format.U32(bytes[membershipRecordIDOffset:membershipRecordWordCountOff])
	locator.wordCount = format.U32(bytes[membershipRecordWordCountOff:membershipRecordLeafPageOff])
	locator.leafPage = format.U32(bytes[membershipRecordLeafPageOff:membershipRecordBlobRootOff])
	locator.leafIndex = format.U16(bytes[membershipRecordLeafIndexOff:membershipRecordIDOffset])
	locator.storage = storage
	locator.blobRoot = root
	locator.rejected = bytes[membershipRecordRejectedOff] != 0
	return locator, nil
}

// membershipIndex is the recovered membership locator table (Rust
// MembershipIndex = IdIndex<MembershipCodec>).
type membershipIndex = idIndex[membershipLocator]

// newMembershipIndex builds the membership locator table over the
// store layout (Rust MembershipIndex::new).
func newMembershipIndex(tables *tableStore) *membershipIndex {
	return newIDIndex(tables, membershipCodec())
}
